[lldb][RISCV] fix LR/SC atomic sequence handling in lldb-server

daniilavdeev opened the pull request:
lldb-server had limited support for single-stepping through the lr/sc atomic sequence. This patch enhances that support for all possible atomic sequences.

JDevlieghere commented:
CC @SEmmmer

DavidSpickett commented:
Please update the title/description to mention that this is within atomic sequences. I thought this was some kind of fused instruction pair/delay slot type deal but it's not that.

Looks fine at a high level, but I always like to get an architecture expert in just in case. @asb you or someone else you know of? Confirming the claims about the format of the sequence would be enough I think.

dlav-sc commented:
I'm going to land this by the end of the week. If you have any concerns, please let me know.

asb commented:
My apologies for missing this before. It is indeed completely valid to have a compressed branch immediately after the lr.w or sc.w and remain compliant with the forward progress guarantee. I feel I'm lacking a bit of higher level understanding of what this code is trying to do it, because:
* It matches only one of many possible lr.[wd]/sc.[wd] sequences. For instance we'll have sequences of other integer instructions between the lr and sc for e.g. narrower than 32-bit cmpxchg (see llvm/test/CodeGen/RISCV/atomic-cmpxchg.ll)
* Why isn't there a a call to m_emu.WriteMem in the implementation?

dlav-sc commented:
> It matches only one of many possible lr.[wd]/sc.[wd] sequences.

I was confused too when I saw this for the first time. I expected to see a processing a sequence of integer instructions between lr and sc. This patch was enough to fix my issue, so I decided to bring it here as is. However, I'm now thinking I could try to do something more comprehensive and support any possible sequence of instructions.

> Why isn't there a a call to m_emu.WriteMem in the implementation?

From what I understand, it's not a real simulator. This is used in lldb for software stepping to determine where to set a breakpoint. lldb needs to emulate the instruction to obtain next PC.

daniilavdeev committed:
[lldb][RISCV] add lr/sc step tests

daniilavdeev committed:
[lldb][RISCV] handle lr/sc single step

lldb-server had limited support for single-stepping through the
lr/sc atomic sequence. This patch enhances that support for all
possible atomic sequences.

dlav-sc commented:
@DavidSpickett, @asb could you take another look at this patch, please?

I've added support for instructional step for any possible riscv atomic sequence (or at least I believe so). Additionally, I've made some minor refactoring to the software's single-step logic.

DavidSpickett commented on lldb/include/lldb/Core/EmulateInstruction.h:
```diff
@@ -32,6 +32,31 @@ class RegisterValue;
+                                            Status &error);
+
+  std::unique_ptr<EmulateInstruction> m_emulator_up;
+  bool m_emulation_result = false;
```
I'll find out later I expect but what does this do? Is it whether it has an emulation result that can be used?

Seems like it's whether the last emulation call succeeded.

DavidSpickett commented on lldb/test/API/riscv/step/TestSoftwareStep.py:
```diff
@@ -0,0 +1,90 @@
+        )
+
+        pc = cur_thread.GetFrameAtIndex(0).GetPC()
+        self.assertTrue((pc - entry_pc) == 0x4)
```
I think the only changing parts of the tests are the source, exe names and the value you check the pc with.

Could be refactored to:
```
def do_sequence_test(....):
  return pc

def test_...
  self.do_sequence_test(...
  self.assertTrue(whatever check you wanted to do on the PC)
```

This also makes the tests a bit easier to read because the PC assert is now right next to the test name so I can see sequence type A checks the PC this way.

dlav-sc replied:
addressed

DavidSpickett reviewed (commented):
If I'm someone looking to find out if a particular atomic sequence is valid, where do I look, what's the ABI / ISA doc that tells us that?

I don't literally want to do that now, but someone in future might, so if there is one, please add links to it in the PR message and the top of the test file.

(if there is no stable link, the doc's name is enough)

daniilavdeev committed:
add comments

DavidSpickett commented on lldb/source/Plugins/Instruction/RISCV/EmulateInstructionRISCV.cpp:
```diff
@@ -1792,4 +1778,127 @@ bool EmulateInstructionRISCV::SupportsThisArch(const ArchSpec &arch) {
+              "RISCVSingleStepBreakpointLocationsPredictor::%s: can't find "
+              "corresponding store conditional insturuction",
+              __FUNCTION__);
+    return {entry_pc + 4};
```
Asking this again because I think GitHub hid the comment, is +4 safe here? Could it be +2 if the instruction is compressed, or do you already know that it won't be?

dlav-sc commented on lldb/include/lldb/Core/EmulateInstruction.h:
```diff
@@ -32,6 +32,31 @@ class RegisterValue;
+  virtual ~SingleStepBreakpointLocationsPredictor() = default;
+
+protected:
+  lldb::addr_t GetSequentiallyNextInstructionPC(Status &error);
```
I mean next in the binary. I've added a comment about it.

dlav-sc commented on lldb/source/Plugins/Instruction/ARM/EmulateInstructionARM.h:
```diff
@@ -16,6 +16,16 @@
+class ARMSingleStepBreakpointLocationsPredictor
+    : public SingleStepBreakpointLocationsPredictor {
+public:
+  ARMSingleStepBreakpointLocationsPredictor(
+      std::unique_ptr<EmulateInstruction> emulator_up)
+      : SingleStepBreakpointLocationsPredictor{std::move(emulator_up)} {}
```
Honestly, I'm not sure what you mean 😅. I need a constructor that takes a `std::unique_ptr<EmulateInstruction>`.

dlav-sc replied:
Nope.. that won't work. It can't deduce a non-default constructor.

dlav-sc commented on lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.h:
```diff
@@ -23,7 +23,7 @@ class NativeProcessSoftwareSingleStep {
   // List of thread ids stepping with a breakpoint with the address of
   // the relevan breakpoint
-  std::map<lldb::tid_t, lldb::addr_t> m_threads_stepping_with_breakpoint;
+  std::multimap<lldb::tid_t, lldb::addr_t> m_threads_stepping_with_breakpoint;
```
To properly handle `lr`/`sc` sequence, `lldb` sometimes needs to set multiple breakpoints. This means there can now be several breakpoints corresponding to the same thread. I use `std::multimap` here because it allows multiple values with the same key, unlike `std::map` which requires unique keys.

DavidSpickett replied:
Sorry yes I'm thinking of a bi-directional map.

dlav-sc commented on lldb/test/API/riscv/step/TestSoftwareStep.py:
```diff
@@ -0,0 +1,90 @@
+"""
+Test software step-inst
```
addressed

dlav-sc commented on lldb/test/API/riscv/step/TestSoftwareStep.py:
```diff
@@ -0,0 +1,90 @@
+        )
+
+        pc = cur_thread.GetFrameAtIndex(0).GetPC()
+        self.assertTrue((pc - entry_pc) > 0x10)
```
From what I understand, lldb runs the same test twice: the first one with the default dwarf debug info embedded in the binary (`dwarf` mode) and the second one with a binary compiled using `-dwarf-split`, which stores debug info in a separate file (`dwo` mode) (I might be mistaken about this exact workflow.)

I actually noticed a while back (about a year or more ago) that `-dwarf-split` doesn't work properly for riscv. `Clang` had been failing during compilation with an error stating that `-dwarf-split` is unsupported for riscv. However, more recently, clang seems to start compile something, but I'm not sure if the output is valid - some `dwo` tests show strange behavior, which might be related to issues with split debug info.

Regarding this specific test, it works as expected with standard `dwarf`, but there are problems with `dwo`:
* The test with incomplete `lr`/`sc` passes with `dwo`
* The branch test behaves strangely (doesn't even stop after step instruction)
* The standard `lr`/`sc` sequence test works but shows different PC offsets (`pc - entry_pc`)

After examining the binary a bit, looks like it uses different relocations than the dwarf version, resulting in 1-2 extra instructions in atomic sequences. This explains the different offsets in the same test but with different debug info formats .

This seems odd to me, as I wouldn't expect `-dwarf-split` to affect code generation. I've been meaning to investigate` -dwarf-split` more thoroughly. 

For now we could temporarily exclude `dwo` tests at all from consideration and use the offsets that corresponds to the standard `dwarf` case.

DavidSpickett commented on lldb/test/API/riscv/step/incomplete_sequence_without_lr.c:
```diff
@@ -0,0 +1,19 @@
+void __attribute__((naked)) incomplete_cas(int *a, int *b) {
```
Actually I think it's not missing, it's doing something else? The sc.w part.

Anyway, a comment will make this clear.

DavidSpickett commented on lldb/test/API/riscv/step/TestSoftwareStep.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class TestSoftwareStep(TestBase):
+    @skipIf(archs=no_match(re.compile("rv*")))
```
I'd just remove the re.compile, it's probably not improving much, and adding some visual noise.

daniilavdeev committed:
add more comments

dlav-sc commented on lldb/source/Plugins/Process/FreeBSD/NativeProcessFreeBSD.cpp:
```diff
@@ -326,11 +326,15 @@ void NativeProcessFreeBSD::MonitorSIGTRAP(lldb::pid_t pid) {
-            LLDB_LOG(log, "pid = {0} remove stepping breakpoint: {1}",
-                     thread_info->first, brkpt_error);
-          m_threads_stepping_with_breakpoint.erase(thread_info);
+          while (thread_info != m_threads_stepping_with_breakpoint.end() {
```
In `NativeProcessLinux`, it simply removes all breakpoints at once using `m_threads_stepping_with_breakpoint.clear()` after stop. I'm not sure why the `FreeBSD` implementation deletes breakpoints one by one instead.

daniilavdeev committed:
replace multimap with map of vectors

DavidSpickett reviewed (approved):
LGTM

slydiman commented:
[lldb-remote-linux-ubuntu](https://lab.llvm.org/buildbot/#/builders/195/builds/10876) is broken after this patch.
Please update the test conditions to skip riscv tests on non-riscv platforms.

JDevlieghere commented:
GreenDragon too: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/28268/

DavidSpickett commented:
Yeah to my surprise it's `re.search` not `re.match`. Makes sense in hindsight though.
